### ISSUE TITLE
ExpressVPN not working

### ISSUE THREAD

**Issue by sasagr** [Zomboided/service.vpn.manager#7]
I think I did all correct but I m getting this error:
-------------------
Thu Nov  3 19:11:50 2016 OpenVPN 2.3.12 armv7ve-libreelec-linux-gnueabi [SSL (OpenSSL)] [LZO] [EPOLL] [IPv6] built on Oct 20 2016
Thu Nov  3 19:11:50 2016 library versions: LibreSSL 2.3.8, LZO 2.09
Thu Nov  3 19:11:50 2016 WARNING: file '/storage/.kodi/userdata/addon_data/service.vpn.manager/ExpressVPN/user.key' is group or others accessible
Thu Nov  3 19:11:50 2016 WARNING: file '/storage/.kodi/addons/service.vpn.manager/ExpressVPN/ta.key' is group or others accessible
Thu Nov  3 19:11:50 2016 Control Channel Authentication: using '/storage/.kodi/addons/service.vpn.manager/ExpressVPN/ta.key' as a OpenVPN static key file
Thu Nov  3 19:11:51 2016 UDPv4 link local: [undef]
Thu Nov  3 19:11:51 2016 UDPv4 link remote: [AF_INET]31.3.155.40:1195
Thu Nov  3 19:11:51 2016 VERIFY ERROR: depth=0, error=format error in certificate's notAfter field: C=VG, ST=BVI, O=ExpressVPN, OU=ExpressVPN, CN=Server-349-1a, emailAddress=support@expressvpn.com
Thu Nov  3 19:11:51 2016 OpenSSL: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Thu Nov  3 19:11:51 2016 TLS_ERROR: BIO read tls_read_plaintext error
Thu Nov  3 19:11:51 2016 TLS Error: TLS object -> incoming plaintext read error
Thu Nov  3 19:11:51 2016 TLS Error: TLS handshake failed
Thu Nov  3 19:11:51 2016 SIGUSR1[soft,tls-error] received, process restarting
Thu Nov  3 19:11:53 2016 UDPv4 link local: [undef]
Thu Nov  3 19:11:53 2016 UDPv4 link remote: [AF_INET]31.3.155.40:1195
Thu Nov  3 19:11:53 2016 VERIFY ERROR: depth=0, error=format error in certificate's notAfter field: C=VG, ST=BVI, O=ExpressVPN, OU=ExpressVPN, CN=Server-349-1a, emailAddress=support@expressvpn.com
Thu Nov  3 19:11:53 2016 OpenSSL: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed
Thu Nov  3 19:11:53 2016 TLS_ERROR: BIO read tls_read_plaintext error
Thu Nov  3 19:11:53 2016 TLS Error: TLS object -> incoming plaintext read error
Thu Nov  3 19:11:53 2016 TLS Error: TLS handshake failed
Thu Nov  3 19:11:53 2016 SIGUSR1[soft,tls-error] received, process restarting
-----------------

Cannot understand where is the issue.

**Comment by Zomboided** [comment]
I suspect you've not created the key and cert files correctly.  I noticed the wiki wasn't displaying correctly because I used tags without escaping them.  I've updated the page https://github.com/Zomboided/service.vpn.manager/wiki/02.-Basic-Usage now, delete the key and cert you have (settings/utilities) and then try connecting again.

**Comment by sasagr** [comment]
Those files have been created correctly. Actually expressvpn provide them separately and I just needed to rename them to user... Otherwise the content is in the format you described in the wiki.
What I m not sure is if login user and pwd are required. In any case in both cases (with and w/o) I get the same error. 
Any other idea?
I tried to use the user defined option but do not know how exactly to set it up and if it really makes sense to try it

**Comment by Zomboided** [comment]
It might be a problem with your network getting connection, but I think it's more likely that they've changed something.  Can you fetch me the latest ovpn files from ExpressVPN and attach them here.

**Comment by sasagr** [comment]
Here it is. I removed the cert and key not to be published here but if you need I will pass he full file to you in private.
I made the exercise to compare the ovpn (yours and mine) and I did not see significative difference, only some buffer value.
To be sure nothing is wrong with my certs I double check the same ovpn with Tunnelblick for MacOSX and I did not have any issue so I can exclude issues with certs.

Pls let me know what you find out and in case you want the full ovpn file pls guide me how to send it to you.
Thanks
[my_expressvpn_sweden_udp.ovpn.zip](https://github.com/Zomboided/service.vpn.manager/files/570830/my_expressvpn_sweden_udp.ovpn.zip)

**Comment by Zomboided** [comment]
So assuming you have well formed keys, as per the wiki:
Access the addon directory (for libreelec this is /storage/.kodi/addons/service.vpn.manager/ExpressVPN)
Edit the TEMPLATE.txt file to remove "#REMOVE1' and '#REMOVE2'
Access the addon's Settings/Utilities screen in Kodi and select the option to reset the ovpn files.
Then try reconnecting.  If this works, then try connecting using a TCP profile as well.
Let me know how it goes and I'll make changes if needed.

**Comment by sasagr** [comment]
It does not work. Here the Log.
I m not sure if it needs login and password to be input but I tried with and without and get same error.

[openvpn.log.zip](https://github.com/Zomboided/service.vpn.manager/files/571503/openvpn.log.zip)

**Comment by Zomboided** [comment]
Hmmm, dunno then.  Password will be ignored anyway as the opvn file doesn't seem to use it (therefore I don't go and look).  If you want to lend me your ID to try and fix it then PM me on the libreelec forums.

**Comment by sasagr** [comment]
I will send you in trust the full ovpn file so you can test it. Where do I find you on libreelec forums? my id there is sasagr

**Comment by sasagr** [comment]
I found you in the libreelec forum with the same name. sent you the file.

**Comment by Zomboided** [comment]
Tested with keys, worked fine.  Check your network and the message I sent you back.  Good luck fixing.

**Comment by elektrrodynamiks** [comment]
Hi - I am having the same problem and I was hoping to find the answer at the end of this thread :(
Could you please hint on a settings for expressvpn. thanks.

**Comment by Zomboided** [comment]
Install the latest version and follow the instructions.  Worked just fine when I used sasagr's account.

**Comment by Schaka** [comment]
I have the same problem with VPNSecure. Keys are selected correctly (I've tried embedding and supplying separately).

But the error remains the same after asking for password and accepting it. The password is correct, as the error message changes if I enter an incorrect password.
Any idea what could cause this?

@sasagr  did you ever solve this?

**Comment by Zomboided** [comment]
https://github.com/Zomboided/service.vpn.manager/issues/48

**Comment by dhayo** [comment]
Hello! Need help please!
About username and password
How can i get that?

**Comment by kaptajnos** [comment]
hey Zom im having the same issue here. cant get it to connect to ekspress . help would surely be nice im almost losing my mind here,

**Comment by Zomboided** [comment]
See the linked issue above. It's a problem with some versions of openvpn, nothing I can do to fix it. Try newer builds

**Comment by kjellmoens** [comment]
Hi, 

I had the same problem with the keys. Deleted my keys and now I get Insufficient key material or header text not found in file '/storage/.kodi/addons/service.vpn.manager/ExpressVPN/ta.key

**Comment by Zomboided** [comment]
If the downloaded ta.key is broken (the one which appears to work for
everyone else), then reset the VPN provider and then remove the downloaded
files using the Settings/Utilities.
You can then reconnect, and supply your private keys (or better, supply an
ovpn and let the add-on get the key for you).
My money is on your keys being incorrect here.  But you've not bothered to
supply any form of log or debug so I don't really know.

**Comment by Kamosh** [comment]
Hi all,
today I was hit by the same ugly error: 
VERIFY ERROR: depth=0, error=format error in certificate's notAfter field: C=VG, ST=BVI, O=ExpressVPN, OU=ExpressVPN, CN=Server

I had OpenELEC version 8.0.4 (or 8.0.5?) on Raspberry PI 2. 
Seems that the openssl library the OpenELEC uses
> /usr/bin/openssl version 
LibreSSL 2.4.5)
which is not able to cope with certificates valid after 2050 (some bugs say 2038). I checked my ExpressVPN certificate and it is valid to: November 2, 2066.

The solution in my case - I migrated from OpenELEC to LibreELEC. LibreELEC uses a different library for ssl:
> /usr/bin/openssl
OpenSSL 1.0.2o  27 Mar 2018

I am now able to establish a VPN connection using VPN Manager for OpenVPN (BTW great plugin) and the ExpressVPN. Migration from OpenELEC to LibreELEC: https://wiki.libreelec.tv/how_to/update_libreelec.

Happy VPNing,
Kamosh

**Comment by skaftisveins** [comment]
I been trying to make this work for the last hours and never made it past the authentication check. My question for starters is simple. Will I need to edit the .ovpn file I downloaded from expressvpn in order for this to work. I see some guides where op add code in the .ovpn file to locate where the .crt and pass.txt are located on the kodi libreelec

**Comment by skaftisveins** [comment]
> No, use the built in expressvpn support (I think it's supported....) and when it asks you for an ovpn file for the key (I think it has a key per connection?) just supply it and it'll extract the key out
> […](#)
> On 23 Jan 2019, at 02:52, skaftisveins ***@***.***> wrote: I been trying to make this work for the last hours and never made it past the authentication check. My question for starters is simple. Will I need to edit the .ovpn file I downloaded from expressvpn in order for this to work. I see some guides where op add code in the .ovpn file to locate where the .crt and pass.txt are located on the kodi libreelec — You are receiving this because you modified the open/close state. Reply to this email directly, view it on GitHub, or mute the thread.

okay I will give it another try, I  am removing the addon and downloading a new copy of .ovpn from ExpressVPN. One thing I noticed though, I am using version 5.0.7 when the newest looks like to be 5.0.8 at least according to addon.xml. When I try to update in kodi it just show me 5.0.7 as the newest version. 

Should I download the 5.0.8 zip and replace the files in the folder where it was installed? 
( /home/osmc/.kodi/addons/service.vpn.manager )

**Comment by Zomboided** [comment]
Good news...the addon does all sorts of messing with the xml, just use the
GUI and you'll be fine.
Cheers

**Comment by skaftisveins** [comment]
Hey sorry I'm running into a problem again. It says connected and everything looks good. Then click, "Display VPN status" and everything is correct in the pop up window. In about 5 minutes or so I try the same and the pop up show it's using .ovpn but everything else is "no info" or "unknown". I tried running DNS fix in the GUI and it did not solve and after trying I was not able to connect the VPN again. Had to flush everything first in advanced settings before I could connect again. Anyway sorry for the annoying issues but I will post the openvpn.log file for you to look at if you have any ideas how to solve, thanks again! (also, feel free to let me know if there is any information the log file that I should be leaving out before posting here)
[openvpn.log](https://github.com/Zomboided/service.vpn.manager/files/2787889/openvpn.log)

**Comment by Zomboided** [comment]
Can you post an expressvpn ovpn file please?  If there are keys in it, just
xxx over them.

**Comment by skaftisveins** [comment]
Of course, will this do ?
[nokeys.zip](https://github.com/Zomboided/service.vpn.manager/files/2788428/nokeys.zip)

**Comment by Zomboided** [comment]
if you go here, or similar depending on where OSMC deposits the userdata
/.kodi/userdata/addon_data/service.vpn.manager/Downloads/expressvpn
you'll see a file called TEMPLATE.txt
Change "comp-lzo" to "comp-lzo no"
Then go to settings/utility, Reset VPN provider
Then re-validate a connection.  Don't reinstall or delete the downloaded
files.
Check a few different connections if it works. This is the only difference
between what you shared and the parameters I was using.

I'm kinda surprised to see the error message you're seeing because of this,
but who knows.    If it doens't work, then you should also check that
you've enabled "Check connection with ping" in settings/advanced before
resetting the provider as above.  It looks like it's disconnecting
repeatedly, which this option is designed to combat.  I'd expect to see it
say that it's disconnected through lack of connectivity though.

**Comment by skaftisveins** [comment]
I will give this a try when I have the chance and post the results here. Thanks man!

**Comment by skaftisveins** [comment]
I followed your instructions but unfortunately to no success. I edited the TEMPLATE.txt and then I verified the "Check connection with ping" was checked in settings. I only Resest VPN provider and then I was not able to make a connection again with these exact settings. After this I flushed everything and started from scratch and I made a change to config in /home/osmc/.config/expressvpn and set auto_connect = false. I think it's stable now. The log is at least not constantly updating with some errors and if I check the VPN status in GUI it stays the same. I think we got this, Yo!

**Comment by Zomboided** [comment]
Ok good...sounds like you had something in .config that was fighting with
the addon.  You don't need to put anything in .config for this, the addon
will do it all for you, and let you connect on boot.

**Comment by mrhfb80** [comment]
I'm a noob with all of this so bear with me. I too am trying to set up ExpressVPN in Kodi 17.6 with VPN Manager 6.0.2. I've reread things multiple times so as to not miss something obvious but my attempts keep failing. Via the wizard I can enter user name and password (doubled checked), it will run through a few steps, I can choose the server location, I select the proper .ovpn file but then every time it fails. I always get, "Error connecting to VPN, connection has timed out or VPN could not be reached, Retry, or try using a different port or VPN profile." 

When rereading the dialogue between you and @skaftisveins I realized he said he was using "kodi libreelec." I however am using OpenELEC. Should the solution you two worked out apply to my situation as well? 

[kodi_19-02-12_14-14-58.log](https://github.com/Zomboided/service.vpn.manager/files/2854352/kodi_19-02-12_14-14-58.log)

**Comment by Zomboided** [comment]
Well you've only provided a partial log, so I'm not sure what's going on.
From what else I do see in the log it looks like you can't even reach the VPN - so there's some network or DNS issue going on even before the VPN connects (ie the VPN is not messing up the network connection)

I think the previous user was about to connect (ie network was fine) before experiencing auth errors and then constant drop outs.

**Comment by mrhfb80** [comment]
My bad, didn't think the whole log was needed. I'm attaching a complete log this time. Is there any other info that I can share to help? Whatever advice you have for troubleshooting the network/DNS issue would be much appreciated.

One thing to note is that I am currently behind the "Great Firewall," I don't know if that would be a factor at play but I thought I'd mention it just in case. I know Kodi can connect to the network, I can download addons, but I do think somethings are blocked. Also, I have no issues with phones, laptops, Raspbian connecting to select Express servers. Thanks for helping me out!

[kodi_19-02-12_14-05-16.log](https://github.com/Zomboided/service.vpn.manager/files/2855088/kodi_19-02-12_14-05-16.log)

**Comment by Zomboided** [comment]
This : 12:57:38.163 T:1905259264  NOTICE: Tue Feb 12 12:56:35 2019 RESOLVE:
Cannot resolve host address: hongkong-3-ca-version-2.expressnetw.com:1195
(Name or service not known)
is an example of your DNS not resolving the host address, so basically
there's not a route to the VPN server.

This : One thing to note is that I am currently behind the "Great
Firewall,"
is why it can't resolve the host address.  You can google VPN and China to
understand the situation, but basically VPNs are blocked (banned?) in China.

Some VPN providers try and provide ways around this by using different
ports and different protocols.  I've no idea if ExpressVPN is one of them,
again, you can google this (or ask them).  If they do offer a workaround,
then I can't tell you if the add-on will support this until I know what it
is.

**Comment by mrhfb80** [comment]
Before I have made the error in selecting a location specific .ovpn file and then chosing a different location during the wizard set up. Would that create this issue? I just gave it another go with both .ovpn and server locations agreeing (Japan - Tokyo - 1). Then when I checked the log I couldn't find the same error message.  

But just to be clear I can access "Japan - Tokyo - 1" on other devices without issue in China because ExpressVPN has made some locations to be China friendly. I'm using it right now on my laptop  and to be sure I just switched to Raspbian and the "Japan - Tokyo - 1" successfully connected. This stuff is way over my head, would Kodi or OpenVPN operate differently therefore encountering a firewall issue that other devices don't currently have?

What should I specifically ask Express about in regards to ports and protocols? 

[kodi_19-02-12_19-10-53.log](https://github.com/Zomboided/service.vpn.manager/files/2855610/kodi_19-02-12_19-10-53.log)

**Comment by Zomboided** [comment]
The friendly locations probably use TCP rather than UDP for the protocol
and a different port.  You can change these within the add-on on the first
settings page.

It's possible that ExpressVPN has changed a bunch of stuff for these
locations and it's not reflected in the add-on.  I don't check every
location for every single provider supported.

**Comment by mrhfb80** [comment]
I chatted with someone from Express and they said the port values are accurate. The location's ovpn filed states "UDP" but I tried TCP anyways, no go. And I don't blame you for not checking every single location for every single provider supported, that would take a lot of time!

**Comment by TNelson860** [comment]
I'm a bit confused here. I'm using OpenELEC 8.0.4 (I think) and when going through the Wizard to setup ExpressVPN, I get to the point where it's asking for a key. I'm not sure where to point it to as nothing was downloaded. And looking at any and all guides for setup, nothing says anything about any keys to use.

**Comment by Rotem71** [comment]
Hello.

I am trying to setup ExpressVPN on the VPN manager (Raspberry pi 2) after I've used it successfully with NordVPN. I follow the instructions with regard to providing the key file / crt file / ovpn file, but I get an error every time.

I attach a screenshot of the log from the TV.

Can you tell me what am I doing wrong please?

Thank you.

![IMG_20200421_192058](https://user-images.githubusercontent.com/64092962/79889660-07e05000-8407-11ea-9153-aef0b25705cf.jpg)

**Comment by Rotem71** [comment]
I've also turned to ExpressVPN's support, and after trying several things they wrote this:

"Our VPN is not yet optimized to work on OpenELEC. We cannot guarantee that it will work on our configuration. However, what I can do is to send a feature request to our developers to work on this so you can set it up. The tutorial will be available on our website once it's working.

In the meantime, we recommend using a VPN-capable router to share your VPN connection with devices that can't run the VPN yet."

Does that mean that VPN manager finally doesn't work with ExpressVPN?

**Comment by Zomboided** [comment]
Well on the LE forums there are people who have independently installed it saying that it works.  And I've borrowed an account and validated it works.  So I think it works.  And honestly LE is "just" another (well crafted) Linux install so I don't know why their support team thinks it doesn't work.

My observation is that the password and userIDs ExpressVPN use are not the most intuitive and I can believe mistakes can easily be made entering them in a Kodi interface.  You look like you're having authentication issues.  So either you're using the wrong keys or the wrong user ID/password.

**Comment by Rotem71** [comment]
I've downloaded their zip file, which contains their .key and .crt files.
I'm sure I've entered the user/pass details correctly, but I can make a text file of them, copied directly from their web site. I've seen somewhere that I can make a pass.txt file that contains the user and the password. Is that correct?
If so - I can do that, but I believe that I've entered these values correctly.
Is there a way to bypass that? i.e. copy the files manually to to sd card or something?

**Comment by tgilbert2003** [comment]
Hi, I realize this thread is a few years old, but I am having a similar issue as the author.  I am using the latest version of LibreELEC which has the newest version of KODI on it, but I can not get ExpressVPN to work.  I am using the correct username and password provided on my account page for ExpressVPN and I have the keys and certificate files that they provided on my account page as well.  I have tried multiple servers in the United States but none of them will let me connect.  I saved the log using the option in the settings menu.  Hope there is something in there you see as to why I can not get this to work.
[kodi_20-04-27_01-38-59.zip](https://github.com/Zomboided/service.vpn.manager/files/4537904/kodi_20-04-27_01-38-59.zip)

**Comment by Zomboided** [comment]
I'm not opening random .zip file.

Multiple people have tried this and it works.  You have to ensure that your device is aligned with the servers being using.  Or you can use the User Defined wizard with the ovpn files that you can download.  If it says auth error, then it probably is.

There are a lot more words on the LibreELEC forum over the past 3/4 pages of threads about this.

**Comment by Rotem71** [comment]
Dear Zomboided,

ExpressVPN is one of the biggest VPN out there, and it seems that it doesn't work with the OpenVPN anymore.
I've tried it once and again with different variations - nothing works.

I'm ready to send you anything you need to help me out here - but saying that it used to work for other isn't helpful.

I ask again: Is there a way to bypass that? i.e. copy the files manually to the sd card or something?

i'm willing to take any effort to fix that.

**Comment by Zomboided** [comment]
@Rotem71 Someone else lent me an account and this works just fine for me.  And I've given you the alternative of User Defined.  You can absolutely copy all the files you need onto a SD card and use ovpenvpn via the command line if User Defined is not working.

Ultimately i'm not interested in spending more time on this when I perceive the issue to be the complexity of the service.

**Comment by tgilbert2003** [comment]
> I'm not opening random .zip file.
> 
> Multiple people have tried this and it works. You have to ensure that your device is aligned with the servers being using. Or you can use the User Defined wizard with the ovpn files that you can download. If it says auth error, then it probably is.
> 
> There are a lot more words on the LibreELEC forum over the past 3/4 pages of threads about this.

Sorry, I figured it was easier than copy and pasting a long log.  It does have the auth_error in there but I used the Wizard and put all the information in that it asked for and loaded in the keys/certificate files that it asked for.  Thanks anyways I guess.

**Comment by peno64** [comment]
If you want to post a log then it is best to do it this way: https://kodi.wiki/view/Log_file
In short, you go to http://paste.kodi.tv/
You past your log in there and save it (Ctrl S)
Then the link changes to something unique and that link you post here.

**Comment by tgilbert2003** [comment]
> If you want to post a log then it is best to do it this way: https://kodi.wiki/view/Log_file
> In short, you go to http://paste.kodi.tv/
> You past your log in there and save it (Ctrl S)
> Then the link changes to something unique and that link you post here.

Ok thanks, I'll try that.

**Comment by tgilbert2003** [comment]
https://paste.kodi.tv/hirimosica.kodi

**Comment by tgilbert2003** [comment]
I think I found another issue....From what I read, the version of OpenVPN (2.6) is causing the error --ns-cert-type is DEPRECATED. Use --remote-cert-tls instead.   Is this the reason why it will not connect to any VPN's?  If so, I am still a noob at some of this and would like to know,  how do I switch to the other command? Do I edit the certificate file from ExpressVPN?

**Comment by tgilbert2003** [comment]
Wow, found the issue.....Very noobish but I did not think that the font used by ExpressVPN would throw me off.....I was using a 1 instead of a lowercase L.   On Kodi, the 1 looked more like the L so I kept using a 1.....Switched everything and now it is connected.   What a noob.  🙄

**Comment by Rotem71** [comment]
I've noticed the font issue immediately and had the same problem with "1" and "l" - it's actually "l" and not "L" (lowercase rather that uppercase).
However, It still does not work for me and ExpressVPN still holds that their software does not work on OpenElec.

**Comment by tgilbert2003** [comment]
> I've noticed the font issue immediately and had the same problem with "1" and "l" - it's actually "l" and not "L" (lowercase rather that uppercase).
> However, It still does not work for me and ExpressVPN still holds that their software does not work on OpenElec.

For me, the lowercase L instead of the 1 worked perfect and it's been working great so far.

**Comment by Rotem71** [comment]
I'm happy for you, tgilbert2003.

**Comment by peno64** [comment]
@Rotem71 

Try to edit /storage/.kodi/userdata/addon_data/service.vpn.manager/settings.xml yourself with any text editor.

Look in the file for:

    <setting id="vpn_username">.</setting>
    <setting id="vpn_username_validated">...</setting>

and put your username or email address you use for your service where the ... are

and search for

    <setting id="vpn_password">...</setting>
    <setting id="vpn_password_validated">...</setting>

and put your password where the ... are

All this data is in plain ascii.

Save this file and restart your system.

Also if you have used strange characters in there for example é or so, try to change that to something without these strange characters and try again.

**Comment by Rotem71** [comment]
Here is my settings.xml file, where I changed the username and password for this post only.
Using the username and password as I use elsewhere - doesn't work.
I've tried to do what peno64 has suggested and copied my username and password to the proper lines at the settings.xml file but the result is the same.

I've changed the name of the settings.xml files to .txt files just for this post.

[Original settings.txt](https://github.com/Zomboided/service.vpn.manager/files/4548207/Original.settings.txt)
[Modified settings.txt](https://github.com/Zomboided/service.vpn.manager/files/4548214/Modified.settings.txt)
